### PR TITLE
Inst literal

### LINKED ISSUE #49: Create basic literal GIL instructions (`IntegerLiteralInst`, `FloatLiteralInst`, `StringLiteralInst`)

### PR DESCRIPTION
**Pull request by yunsecode**
Closes #49

### CODE REVIEW

File: include/GIL/InstBase.hpp [hunk 1]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
     }
 };
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:429]: question is, do we want to put all instructions in the same file? (Swift does put everything in one file)

File: include/GIL/InstBase.hpp [hunk 2]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
...
+/// @brief A class representing a literal instruction in the GIL.
+///
+/// These instructions are used to control the flow of execution in a function.
+/// They have no results and are always the last instruction in a basic block.
+class ConstantInst : public InstBase {
+public:
+    size_t getResultCount() const override { return 1; }
+    size_t getOperandCount() const override { return 2; }
+
+    virtual Type getResultType(size_t index) const = 0;
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:441]: already virtual pure in InstBase so no need

File: include/GIL/InstBase.hpp [hunk 3]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
...
+protected:
+    Type destType;
+    llvm::APInt literalInt;
+
+public:
+    void setDestType(Type type) { destType = type; }
+    Type getDestType() const { return destType; }
+
+    void setLiteralInt(llvm::APInt value) { literalInt = value; }
+    llvm::APInt getLiteralInt() const { return literalInt; }
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:460]: maybe better names? `type` and `value`?

File: include/GIL/InstBase.hpp [hunk 4]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
...
+    {
+        switch (index) {
+        case 0: return destType;
+        case 1: return literalInt;
+        default: llvm_unreachable("Invalid operand index");
+        }
+    }
+    Type getResultType(size_t index) const override
+    {
+        assert(false && "Result index out of range");
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:472]: return destType;

File: include/GIL/InstBase.hpp [hunk 5]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
...
+            && inst->getKind() <= InstKind::ConstantInstFirstKind;
+    }
+};
+
+class IntegerLiteralInst : public ConstantInst {
+protected:
+    Type destType;
+    llvm::APInt literalInt;
+
+public:
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:455]: add a constructor (taking both values)

File: include/GIL/InstBase.hpp [hunk 6]
```diff
@@ -428,6 +428,118 @@ class TerminatorInst : public InstBase {
...
+    Type getDestType() const { return destType; }
+
+    void setLiteralString(std::string value) { literalString = value; }
+    llvm::StringRef getLiteralString() const { return literalString; }
+
+    Operand getOperand(size_t index) const override
+    {
+        switch (index) {
+        case 0: return destType;
+        case 1: return Operand(llvm::StringRef(literalString));
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:528]: return getLiteralString(); should do the same

File: include/GIL/InstBase.hpp [hunk 7]
```diff
@@ -428,6 +428,128 @@ class TerminatorInst : public InstBase {
...
+    std::string value;
+
+public:
+    StringLiteralInst(Type type, llvm::APInt value)
+        : type(type), value(value) {}
+
+    void setType(Type type) { this->type = type; }
+    Type getType() const { return type; }
+
+    void setValue(std::string value) { this->value = value; }
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:531]: std::move

File: include/GIL/InstBase.hpp (hunk 2 above)
Comment (Snowy1803) [reply include/GIL/InstBase.hpp]: you can juste remove that line

**Review by Snowy1803 — APPROVED** [review]

File: include/GIL/InstBase.hpp [hunk 8]
```diff
@@ -428,6 +428,126 @@ class TerminatorInst : public InstBase {
...
+/// They have no results and are always the last instruction in a basic block.
+class ConstantInst : public InstBase {
+public:
+    size_t getResultCount() const override { return 1; }
+    size_t getOperandCount() const override { return 2; }
+
+    static bool classof(InstBase const *inst)
+    {
+        return inst->getKind() >= InstKind::ConstantInstFirstKind
+            && inst->getKind() <= InstKind::ConstantInstFirstKind;
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:444]: LastKind

File: include/GIL/InstBase.hpp [hunk 9]
```diff
@@ -428,6 +428,126 @@ class TerminatorInst : public InstBase {
...
+    {
+        switch (index) {
+        case 0: return getType();
+        case 1: return getValue();
+        default: llvm_unreachable("Invalid operand index");
+        }
+    }
+    Type getResultType(size_t index) const override
+    {
+        assert(false && "Result index out of range");
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:508]: return type;

File: include/GIL/InstBase.hpp [hunk 10]
```diff
@@ -428,6 +428,126 @@ class TerminatorInst : public InstBase {
...
```
Comment (Snowy1803) [inline include/GIL/InstBase.hpp:542]: pareil

Comment (yunsecode) [reply include/GIL/InstBase.hpp:542]: sorry ? XD

**Review by Snowy1803 — APPROVED** [review]

**Review by jeremyelalouf — APPROVED** [review]

**Pull request merged**